File: Algorithms/Heap/MaxHeap.py
import sys

class MaxHeap:

    def __init__(self, size):
        
        self.heapSize = size + 1
        self.realSize = 0
        self.maxheap = [0] * self.heapSize

    def add(self, value):
        self.realSize += 1
        
        if self.realSize > self.heapSize:
            print("element is full!")
            self.realSize -= 1
            return

        self.maxheap[self.realSize] = value

        index = self.realSize
        parent = self.realSize // 2

        while parent > 0 and self.maxheap[index] > self.maxheap[parent]:
            self.maxheap[index], self.maxheap[parent] = self.maxheap[parent], self.maxheap[index]
            index = parent
            parent = index // 2

    def peek(self) -> int:
        return self.maxheap[1]

    def pop(self) -> int:

        if self.realSize < 1:
            print("Don't have any element!")
            return -sys.maxsize
        else:
            pop_value = self.maxheap[1]
            self.maxheap[1] = self.maxheap[self.realSize]
            self.realSize -= 1

            index = 1
            while index <= self.realSize // 2:
                left_child = index * 2
                right_child = index * 2 + 1

                if self.maxheap[left_child] > self.maxheap[index] or self.maxheap[right_child] > self.maxheap[index]:
                    if self.maxheap[left_child] > self.maxheap[right_child]:
                        self.maxheap[left_child], self.maxheap[index] = self.maxheap[index], self.maxheap[left_child]
                        index = left_child
                    else:
                        self.maxheap[right_child], self.maxheap[index] = self.maxheap[index], self.maxheap[right_child]
                        index = right_child
                else:
                    break

                # if self.maxheap[left_child] > self.maxheap[index]:
                #     self.maxheap[left_child], self.maxheap[index] = self.maxheap[index], self.maxheap[left_child]
                #     index = left_child
                #     continue
                # elif self.maxheap[right_child] > self.maxheap[index]:
                #     self.maxheap[right_child], self.maxheap[index] = self.maxheap[index], self.maxheap[right_child]
                #     index = right_child
                #     continue
                # else:
                #     break
        
            return pop_value

    def size(self) -> int:
        return self.realSize

    def __str__(self) -> str:
        return str(self.maxheap[1 : self.realSize + 1])

if __name__ == "__main__":
    print("MaxHeap...")

    maxHeap = MaxHeap(5)

    maxHeap.add(1)
    maxHeap.add(2)
    maxHeap.add(3)
    # [3,1,2]
    print(maxHeap)
    # 3
    print(maxHeap.peek())
    # 3
    print(maxHeap.pop())
    # 2
    print(maxHeap.pop())
    # 1
    print(maxHeap.pop())
    maxHeap.add(4)
    maxHeap.add(5)
    # [5,4]
    print(maxHeap)

File: Algorithms/Heap/MaxHeap.java
import java.text.NumberFormat.Style;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.*;

public class MaxHeap {

    private int heapSize, realSize;
    private int[] maxHeap;
    
    public MaxHeap(int size) {
        this.heapSize = size;
        this.realSize = 0;
        this.maxHeap = new int[this.heapSize];
    }

    private void add(int value) {
        this.realSize++;

        if (this.realSize > this.heapSize) {
            System.out.println("element is full !!");
            this.realSize--;
            return;
        }

        this.maxHeap[this.realSize] = value;
        
        int index = this.realSize;
        int parent = this.realSize / 2;

        while (parent > 0 && this.maxHeap[parent] < this.maxHeap[index]) {
            int tmp = this.maxHeap[index];
            this.maxHeap[index] = this.maxHeap[parent];
            this.maxHeap[parent] = tmp;
            index = parent;
            parent = index / 2;
        }
    }

    public Integer peek() {
        return this.maxHeap[1];
    }

    public Integer pop() {

        if (this.realSize < 1) {
            System.out.println("Don't have any element !!");
            return Integer.MAX_VALUE;
        } else {
            Integer pop_value = this.maxHeap[1];
            this.maxHeap[1] = this.maxHeap[this.realSize];
            this.maxHeap[this.realSize] = 0;
            this.realSize--;

            int index = 1;
            while (index <= this.realSize / 2) {
                int left_child = index * 2;
                int right_child = index * 2 + 1;

                if (this.maxHeap[left_child] > this.maxHeap[index] || this.maxHeap[right_child] > this.maxHeap[index]) {
                    if (this.maxHeap[left_child] > this.maxHeap[right_child]) {
                        int tmp = this.maxHeap[left_child];
                        this.maxHeap[left_child] = this.maxHeap[index];
                        this.maxHeap[index] = tmp;
                        index = left_child;
                    } else {
                        int tmp = this.maxHeap[right_child];
                        this.maxHeap[right_child] = this.maxHeap[index];
                        this.maxHeap[index] = tmp;
                        index = right_child;
                    }
                } else {
                    break;
                }
            }
            return pop_value;
        }
    }

    public int size() {
        return this.realSize;
    }

    public void print() {
        System.out.println("print ...");
        for (int i = 0; i < this.realSize; i++) {
            System.out.println(this.maxHeap[i]);
        }
    }

    public static void main(String []arg) {
        // MaxHeap maxHeap = new MaxHeap(10);

        // maxHeap.add(10);
        // maxHeap.add(200);
        // maxHeap.add(2000);
        // maxHeap.print();
        // maxHeap.add(100);
        // maxHeap.pop();
        // maxHeap.print();


        Map<String, String> bracketMap = new HashMap<>();
        bracketMap.put(")", "(");
        bracketMap.put("]", "[");
        bracketMap.put("}", "{");

        List<String> t = new ArrayList<>();

        String s = "(())";
        for (int i = 0; i < s.length(); i++) {
            if (bracketMap.containsKey(Character.toString(s.charAt(i)))) {
                if (t.get(t.size() - 1) != bracketMap.get(Character.toString(s.charAt(i)))) {
                    System.out.println("false");
                    // return false;
                } else {
                    t.remove(t.size() - 1);
                }
            } else {
                t.add(Character.toString(s.charAt(i)));
            }
        }
        System.out.println("true");
    }
}
